packages/amplify-auth-sdk/src/stores/secure-store.js

```diff
@@ -19,19 +26,68 @@ export default class SecureStore extends FileStore {
 	/**
 	 * Loads the `keytar` library and initializes the token file.
 	 *
-	 * @param {Object} [opts] - Various options.
+	 * @param {Object} opts - Various options.
+	 * @param {String} opts.homeDir - The path to the home directory containing the `lib`
+	 * directory where `keytar` is located.
 	 * @param {String} [opts.secureServiceName="Axway AMPLIFY Auth"] - The name of the consumer
      * using this library.
 	 * @access public
 	 */
 	constructor(opts = {}) {
+		if (!opts || typeof opts !== 'object') {
+			throw E.INVALID_ARGUMENT('Expected opts to be an object');
+		}
+
+		const { homeDir } = opts;
+		if (!homeDir || typeof homeDir !== 'string') {
+			throw E.INVALID_PARAMETER('Secure store requires the home directory to be specified');
+		}
+
+		const supportedVersion = fs.readJsonSync(path.resolve(__dirname, '..', '..', 'package.json')).keytar;
+		const prefix = path.join(homeDir, 'lib', 'keytar', `${process.platform}_${process.arch}_${process.versions.modules}`);
+		const keytarPath = path.join(prefix, 'node_modules', 'keytar');
+		const pkgJsonFile = path.join(keytarPath, 'package.json');
 		let keytar;
 
 		try {
-			keytar = require('keytar');
+			const installedVersion = fs.readJsonSync(pkgJsonFile).version;
+			if (!semver.satisfies(installedVersion, supportedVersion)) {
+				log(`Installed keytar out-of-date: ${highlight(installedVersion)}, required ${highlight(supportedVersion)}`);
+				throw new Error();
+			}
+
+			log(`Loading keytar: ${highlight(keytarPath)}`);
+			keytar = require(keytarPath);
 		} catch (e) {
-			/* istanbul ignore next */
-			throw E.KEYTAR_NOT_FOUND('"keytar" package not found, is it installed?');
+			const args = [ 'install', `keytar@${supportedVersion}`, '--no-save', '--production', '--prefix', prefix ];
+			if (e.message) {
+				log(`keytar not found, running: ${highlight(`npm ${args.join(' ')}`)}`);
+			}
+			fs.removeSync(keytarPath);
+
+			const env = Object.assign({ NO_UPDATE_NOTIFIER: 1 }, process.env);
+			const result = spawnSync('npm', args, { env, shell: true });
+			log(`npm install exited (code ${result.status})`);
+
+			if (result.error) {
+				log(result.error);
+				if (result.error.code === 'ENOENT') {
```

Comment:
It looks like using `shell: true` makes us lose out on the error prop you're checking here, we might need to drop that and do the whole `cmd.exe /s /c` thing or use something like `cross-spawn` that I think handles things a little better than plain old `spawnSync`

Reply:
Removed `shell` since I don't think we need it. Instead, I added `windowsHide` instead.

Reply:
It's definitely needed to spawn `npm` on Windows, otherwise it's an ENOENT as it doesn't respect PATHEXT (which is what `shell: true` does). To spawn `npm` on Windows we need to either:

* run `npm.cmd` instead
* add `shell: true`, or spawn like `spawnSync('cmd', ['/s', '/c', 'npm', ...otherArgs)`
    * We'd need to ensure we quote `prefix` just incase it contains spaces as `shell: true` doesn't wrap arguments in quotes like `shell: false` does (i.e `spawnSync('mkdir', [ 'foo bar' ], {shell: true });` on Windows will become `mkdir foo bar` not `mkdir "foo bar"`.
* Use something like `cross-spawn` that handles it all